### PR TITLE
BRITECH-176 Added post endpoint to create new sessions

### PR DESCRIPTION
**Pull request by dwybourn**

### CODE REVIEW

File: Sessions/ControllerConstants.fs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+﻿module Sessions.ControllerConstants
+
+[<Literal>]
+let UNPROCESSABLE_ENTITY = 422
```
Comment (jebbinSCL) [inline Sessions/ControllerConstants.fs]: There is a discussion on this on the web that I was just reading. the 422 error code isn't a standard one, but an extension error code for Web Distributed Authoring and Versioning (WebDAV). I think that is part of the reason it isn't available as part of the status codes enum. I was thinking that I prefer to use the 400 error code instead (Bad request) and put the error detail in the message, like we will anyway?

File: Sessions/Sessions.fsproj [hunk 2]
```diff
@@ -1,109 +1,114 @@
...
-      <HintPath>..\packages\Microsoft.Owin.Hosting.2.0.2\lib\net45\Microsoft.Owin.Hosting.dll</HintPath>
-    </Reference>
-    <Reference Include="Microsoft.Owin.Host.HttpListener">
-      <HintPath>..\packages\Microsoft.Owin.Host.HttpListener.2.0.2\lib\net45\Microsoft.Owin.Host.HttpListener.dll</HintPath>
-    </Reference>
-    <Reference Include="System.Configuration" />
-    <Reference Include="System.Xml.Linq" />
-  </ItemGroup>
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (jebbinSCL) [inline Sessions/Sessions.fsproj]: Is there a reason the sessions and packages are marked as a complete replacement? was there a line ending issue?

File: Sessions/SessionsController.fs [hunk 3]
```diff
@@ -20,4 +23,13 @@ type SessionsController() =
...
         | Some session -> x.Request.CreateResponse(session)
-        | None -> x.Request.CreateResponse(HttpStatusCode.NotFound)
\ No newline at end of file
+        | None -> x.Request.CreateResponse(HttpStatusCode.NotFound)
+
+    member x.Post([<FromBody>] newSessionDetail : SessionDetail) = 
+        Log.Information("Received POST request for new session: {@SessionDetail}", newSessionDetail)
+        try
+            createSession newSessionDetail
+            x.Request.CreateResponse(HttpStatusCode.Created)
```
Comment (lclarksl) [inline Sessions/SessionsController.fs]: I think (?) it's pretty standard to return the link to a new resource in the location header of the response, as well as a 201.

Comment (jebbinSCL) [inline Sessions/SessionsController.fs]: I think it should be the id instead of the link. It's easier to create the link from the id if you need it, rather than getting the id out of the link. The specs for the profiles specify the link, and the spec for sessions should do as well, if it doesn't already

File: Sessions/ControllerConstants.fs (hunk 1 above)
Comment (dwybourn) [inline Sessions/ControllerConstants.fs]: I know it's not technically standard, but it's heavily used. A 400 should only be used where the format of the request is wrong. We'd need to update the api docs if we switch to 400. @ianlovell what do you think?

File: Sessions/SessionsController.fs (hunk 3 above)
Comment (lclarksl) [inline Sessions/SessionsController.fs]: I think if it's in the location header then it should be a link - I don't think it makes sense for it to just be an id? An id on its own should be in the body.

File: Sessions/Repositories.fs [hunk 4]
```diff
@@ -44,3 +45,16 @@ let getSession (id: Guid) =
         None
     else
         Some (entityToSessionDetail (Seq.head sessions))
+
+let createSession (sessionsDetail : SessionDetail) = 
+    let commandStr = "insert into sessions(id, title, status, speakerId, adminId, threadId) values
```
Comment (jebbinSCL) [inline Sessions/Repositories.fs]: I think this should be done as apart of a transaction. Just need make sure that System.Transactions is referenced, and call:
use transaction = connection.BeginTransaction()
and 
transaction.Commit()

File: Sessions/SessionsController.fs (hunk 3 above)
Comment (jebbinSCL) [inline Sessions/SessionsController.fs]: Ah yeah, sorry, I meant there should be the id in the body.

File: Sessions/ControllerConstants.fs (hunk 1 above)
Comment (ianlovell) [inline Sessions/ControllerConstants.fs]: The services should return both depending on the type of error - 400 for syntax errors, 422 for semantic errors. So if the request is malformed, e.g. a required property is missing or the property type is invalid, then return 400. 

If the request is well formed but is still unprocessable, then return a 422. So for example if an email property is not a valid email, or if you send an empty string for a required property. 

The error responses are based from the GitHub API, and they describe their client errors at https://developer.github.com/v3/#client-errors.

File: Sessions/Sessions.fsproj (hunk 2 above)
Comment (dwybourn) [inline Sessions/Sessions.fsproj]: I don't think so. I have no idea what happened here.

File: Sessions/ControllerConstants.fs (hunk 1 above)
Comment (dwybourn) [inline Sessions/ControllerConstants.fs]: Changed to 400 with error message.

File: Sessions/SessionsController.fs (hunk 3 above)
Comment (dwybourn) [inline Sessions/SessionsController.fs]: Now returning the id in the body.

File: Sessions/Repositories.fs (hunk 4 above)
Comment (dwybourn) [inline Sessions/Repositories.fs]: Now execution the command inside a transaction.

File: Sessions/Entities.fs [hunk 5]
```diff
@@ -1,6 +1,8 @@
 ﻿namespace Sessions.Entities
 open System
 
+open System
```
Comment (csscottlogic) [inline Sessions/Entities.fs]: That looks like a duplicate?

Comment (dwybourn) [inline Sessions/Entities.fs]: Well spotted.

File: Sessions/Models.fs [hunk 6]
```diff
@@ -1,11 +1,15 @@
 ﻿namespace Sessions.Models
 
 open System
+open Newtonsoft.Json
 
+[<CLIMutableAttribute>]
```
Comment (csscottlogic) [inline Sessions/Models.fs]: That looks like a better option than having a class with explicitly declared automatic properties. Looks like we have to change CorrespondenceItem in the Comms service to match?

File: Sessions/ControllerConstants.fs (hunk 1 above)
Comment (dwybourn) [inline Sessions/ControllerConstants.fs]: There is one issue with our the controllers. They can't return an error message if using a non standard error code. To return 422, we use CreateResponse(422). However, the method that takes in an error message as a parameter has to use a HttpStatusCode. I tried casting 422 to HttpStatusCode which works in c#, but not f#.

File: Sessions/Repositories.fs [hunk 7]
```diff
@@ -44,3 +45,23 @@ let getSession (id: Guid) =
         None
     else
         Some (entityToSessionDetail (Seq.head sessions))
+
+let executeCommandInTransaction (command : MySqlCommand) = 
+    use transaction = connection.BeginTransaction()
+    command.ExecuteNonQuery() |> ignore
+    transaction.Commit()
```
Comment (csscottlogic) [inline Sessions/Repositories.fs]: The docs say this returns void?
https://msdn.microsoft.com/en-us/library/system.data.sqlclient.sqltransaction.commit(v=vs.110).aspx
When it's called below, ...
Actually, hold on. It isn't called below. I'll write a separate comment for that.

File: Sessions/Repositories.fs [hunk 8]
```diff
@@ -44,3 +45,23 @@ let getSession (id: Guid) =
...
+        (@id, @title, @status, @speakerId, @adminId, @threadId)"
+    let command = new MySqlCommand(commandStr, connection)
+    let id = Guid.NewGuid()
+    command.Parameters.Add("@id", MySqlDbType.Guid).Value <- id
+    command.Parameters.Add("@title", MySqlDbType.String).Value <- sessionsDetail.Title
+    command.Parameters.Add("@status", MySqlDbType.Enum).Value <- sessionsDetail.Status
+    command.Parameters.Add("@speakerId", MySqlDbType.Guid).Value <- sessionsDetail.SpeakerId
+    command.Parameters.Add("@adminId", MySqlDbType.Guid).Value <- sessionsDetail.AdminId
+    command.Parameters.Add("@threadId", MySqlDbType.Guid).Value <- sessionsDetail.ThreadId
+    executeCommandInTransaction |> ignore
```
Comment (csscottlogic) [inline Sessions/Repositories.fs]: This isn't being called. It's a function value that is just being ignored, so this line doesn't do anything.

Comment (csscottlogic) [inline Sessions/Repositories.fs]: Do we need ```executeCommandInTransaction command```?

File: Sessions/Repositories.fs [hunk 9]
```diff
@@ -44,3 +45,23 @@ let getSession (id: Guid) =
         None
     else
         Some (entityToSessionDetail (Seq.head sessions))
+
+let executeCommandInTransaction (command : MySqlCommand) =
```
Comment (csscottlogic) [inline Sessions/Repositories.fs]: Generally, repeating a parameter name in the function name stops things reading fluently, e.g. ```command |> executeInTransaction```.

File: Sessions/Repositories.fs (hunk 8 above)
Comment (dwybourn) [inline Sessions/Repositories.fs]: We do. Why was that not a compile error?

Comment (csscottlogic) [inline Sessions/Repositories.fs]: Unit testing would catch that this isn't executing.

Comment (dwybourn) [inline Sessions/Repositories.fs]: I thought we said we were going to hold off unit testing til next sprint?

File: Sessions/Repositories.fs (hunk 9 above)
Comment (dwybourn) [inline Sessions/Repositories.fs]: Renamed.

File: Sessions/Repositories.fs (hunk 8 above)
Comment (jebbinSCL) [inline Sessions/Repositories.fs]: Yup, until next sprint. 
This isn't a compile error because what you are doing is valid. A function is just a value, and the ignore is meant to throw away a value. The simplest way to make this give a compiler error is to make the function return a value from the command (Which I think would be 1 or zero) Another thing could be to make a computational expression out of it, but that will take a bit more thinking and work, so not necessary now.

**Comment by jebbinSCL** [comment]
:shipit:

File: Sessions/Repositories.fs [hunk 10]
```diff
@@ -44,3 +45,23 @@ let getSession (id: Guid) =
...
+
+let executeInTransaction (command : MySqlCommand) = 
+    use transaction = connection.BeginTransaction()
+    command.ExecuteNonQuery() |> ignore
+    transaction.Commit()
+
+let createSession (sessionsDetail : SessionDetail) = 
+    let commandStr = "insert into sessions(id, title, status, speakerId, adminId, threadId) values 
+        (@id, @title, @status, @speakerId, @adminId, @threadId)"
+    let command = new MySqlCommand(commandStr, connection)
```
Comment (jebbinSCL) [inline Sessions/Repositories.fs]: Just realised that I forgot to point something out, this is using the mysqlcommand directly, but since we are using dapper, we don't need to use it and supply all the parameters, we just need to supply our entity. 

Its up to you if you want to do this now, or I can make the change when i do a bit of refactoring after my next pull request.

File: Sessions/Repositories.fs [hunk 11]
```diff
@@ -44,3 +45,23 @@ let getSession (id: Guid) =
...
+    let id = Guid.NewGuid()
+    command.Parameters.Add("@id", MySqlDbType.Guid).Value <- id
+    command.Parameters.Add("@title", MySqlDbType.String).Value <- sessionsDetail.Title
+    command.Parameters.Add("@status", MySqlDbType.Enum).Value <- sessionsDetail.Status
+    command.Parameters.Add("@speakerId", MySqlDbType.Guid).Value <- sessionsDetail.SpeakerId
+    command.Parameters.Add("@adminId", MySqlDbType.Guid).Value <- sessionsDetail.AdminId
+    command.Parameters.Add("@threadId", MySqlDbType.Guid).Value <- sessionsDetail.ThreadId
+    command |> executeInTransaction |> ignore
```
Comment (csscottlogic) [inline Sessions/Repositories.fs]: You don't need ```|> ignore```. If you hover over ```executeInTransaction``` in Visual Studio, you'll see it returns unit.

File: Sessions/Repositories.fs [hunk 12]
```diff
@@ -44,3 +45,23 @@ let getSession (id: Guid) =
         None
     else
         Some (entityToSessionDetail (Seq.head sessions))
+
+let executeInTransaction (command : MySqlCommand) = 
+    use transaction = connection.BeginTransaction()
+    command.ExecuteNonQuery() |> ignore
+    transaction.Commit()
```
Comment (csscottlogic) [inline Sessions/Repositories.fs]: This returns unit, meaning the whole function returns unit. (See below - you don't need ```|> ignore``` since nothing is being returned from this function.)

File: Sessions/SessionsController.fs [hunk 13]
```diff
@@ -20,4 +22,13 @@ type SessionsController() =
...
\ No newline at end of file
+        | None -> x.Request.CreateResponse(HttpStatusCode.NotFound)
+
+    member x.Post([<FromBody>] newSessionDetail : SessionDetail) = 
+        Log.Information("Received POST request for new session: {@SessionDetail}", newSessionDetail)
+        try
+            let sessionId = createSession newSessionDetail
+            x.Request.CreateResponse(HttpStatusCode.Created, sessionId)
+        with
+            | :? SqlException as ex-> x.Request.CreateResponse(HttpStatusCode.BadRequest, ex.Message)
```
Comment (csscottlogic) [inline Sessions/SessionsController.fs]: Do we know why this might happen? "Bad request" seems quite general.

File: Sessions/SessionsController.fs [hunk 14]
```diff
@@ -20,4 +22,13 @@ type SessionsController() =
...
+            | _ -> x.Request.CreateResponse(HttpStatusCode.InternalServerError)
```
Comment (csscottlogic) [inline Sessions/SessionsController.fs]: This does seem appropriate though!

File: Sessions/Repositories.fs (hunk 10 above)
Comment (dwybourn) [inline Sessions/Repositories.fs]: I've changed the code to use dapper now.

File: Sessions/Repositories.fs [hunk 15]
```diff
@@ -44,3 +44,18 @@ let getSession (id: Guid) =
         None
     else
         Some (entityToSessionDetail (Seq.head sessions))
+
+let createSession (sessionDetail : SessionDetail) = 
+    let args = { Id = Guid.NewGuid()
```
Comment (jebbinSCL) [inline Sessions/Repositories.fs]: It looks like this args variable isn't needed or isn't what you mean? sessionDetail and args are both of type SessionDetail. I'm guessing that you want to set the Id with the new Guid? You can shorten this to: 
`let args = {sessionDetail with Id = Guid.NewGuid() }`

**Comment by jebbinSCL** [comment]
:shipit: after args change

File: Sessions/SessionsController.fs (hunk 13 above)
Comment (dwybourn) [inline Sessions/SessionsController.fs]: An SqlException is quite vague, it could be a huge number of things, we'd have to read the error message for more information. That's why we give the error message in the response.

File: Sessions/Repositories.fs (hunk 15 above)
Comment (dwybourn) [inline Sessions/Repositories.fs]: Thanks, I was googling around for something like this.

File: Sessions/SessionsController.fs [hunk 16]
```diff
@@ -20,4 +22,13 @@ type SessionsController() =
         let session = getSession id
         match session with
         | Some session -> x.Request.CreateResponse(session)
-        | None -> x.Request.CreateResponse(HttpStatusCode.NotFound)
\ No newline at end of file
+        | None -> x.Request.CreateResponse(HttpStatusCode.NotFound)
+
+    member x.Post([<FromBody>] newSessionDetail : SessionDetail) =
```
Comment (lclarksl) [inline Sessions/SessionsController.fs]: This is super picky, but do we have a standard for specifying types? Elsewhere is this repo we have newSessionDetail: SessionDetail

Comment (dwybourn) [inline Sessions/SessionsController.fs]: I'm not sure what you mean? Do you mean no space after the newSessionDetail?

Comment (lclarksl) [inline Sessions/SessionsController.fs]: Yeah

Comment (jebbinSCL) [inline Sessions/SessionsController.fs]: Ah, just remember to use the formatting tool. There are some things it does that I don't like, and I personally change, but it will deal with most things well. Ctrl-(KD)

Comment (dwybourn) [inline Sessions/SessionsController.fs]: What exactly don't you like? If we're using a formatting tool we should all be using the same settings? Maybe we can agree on formatting and push the settings as part of the infrastructure project.

Comment (dwybourn) [inline Sessions/SessionsController.fs]: I won't make any changes just now, as the formatting tools leaves the code in question as is.

Comment (lclarksl) [inline Sessions/SessionsController.fs]: I don't mind what we do as long as it's consistent. For example in Repositories.fs, we have no space before the colon.

Comment (jebbinSCL) [inline Sessions/SessionsController.fs]: I don't think stores it to a shareable file. Just use the defaults for now. It was just some minor stuff about where it moves certain things.

Comment (lclarksl) [inline Sessions/SessionsController.fs]: I am being super picky by the way but I just thought I'd mention it

Comment (jebbinSCL) [inline Sessions/SessionsController.fs]: Want to just run the formatting on Repositories and see if it changes anything? Lets merge after that.

Comment (dwybourn) [inline Sessions/SessionsController.fs]: Done.

**Comment by lclarksl** [comment]
:shipit:
